drivers/i2c/i2c_cc32xx.c

```diff
@@ -0,0 +1,395 @@
+
+	return 0;
+}
+
+static void _i2c_cc32xx_prime_transfer(struct device *dev, struct i2c_msg *msg,
+				      u16_t addr)
+{
+	struct i2c_cc32xx_data *data = DEV_DATA(dev);
+	u32_t base = DEV_BASE(dev);
+
+	/* Initialize internal counters and buf pointers: */
+	data->msg = *msg;
+	data->slave_addr = addr;
+
+	/* Start transfer in Transmit mode */
+	if (IS_I2C_MSG_WRITE(data->msg.flags)) {
+
+		/* Specify the I2C slave address */
+		MAP_I2CMasterSlaveAddrSet(base, addr, false);
+
+		/* Update the I2C state */
+		data->state = I2C_CC32XX_WRITE_MODE;
+
+		/* Write data contents into data register */
+		MAP_I2CMasterDataPut(base, *((data->msg.buf)++));
+
+		/* Start the I2C transfer in master transmit mode */
+		MAP_I2CMasterControl(base, I2C_MASTER_CMD_BURST_SEND_START);
+
+	} else {
+		/* Start transfer in Receive mode */
+		/* Specify the I2C slave address */
+		MAP_I2CMasterSlaveAddrSet(base, addr, true);
+
+		/* Update the I2C mode */
+		data->state = I2C_CC32XX_READ_MODE;
+
+		if (data->msg.len < 2) {
+			/* Start the I2C transfer in master receive mode */
+			MAP_I2CMasterControl(base,
+				       I2C_MASTER_CMD_BURST_RECEIVE_START_NACK);
+		} else {
+			/* Start the I2C transfer in burst receive mode */
+			MAP_I2CMasterControl(base,
+					    I2C_MASTER_CMD_BURST_RECEIVE_START);
+		}
+	}
+}
+
+static int i2c_cc32xx_transfer(struct device *dev, struct i2c_msg *msgs,
+			       u8_t num_msgs, u16_t addr)
+{
+	struct i2c_cc32xx_data *data = DEV_DATA(dev);
+	int retval = 0;
+
+	/* Acquire the driver mutex */
+	k_sem_take(&data->mutex, K_FOREVER);
+
+	/* Iterate over all the messages */
+	for (int i = 0; i < num_msgs; i++) {
+
+		/* Begin the transfer */
+		_i2c_cc32xx_prime_transfer(dev, msgs, addr);
+
+		/* Wait for the transfer to complete */
+		k_sem_take(&data->transfer_complete, K_FOREVER);
+
+		/* Return an error if the transfer didn't complete */
+		if (data->state == I2C_CC32XX_ERROR) {
+			retval = -EIO;
+			break;
+		}
+
+		/* Move to the next message */
+		msgs++;
+	}
+
+	/* Release the mutex */
+	k_sem_give(&data->mutex);
+
+	return retval;
+}
+
+static void _i2c_cc32xx_isr_handle_write(u32_t base,
+					 struct i2c_cc32xx_data *data)
+{
+	/* Decrement write Counter */
+	data->msg.len--;
+
+	/* Check if more data needs to be sent */
+	if (data->msg.len) {
+
+		/* Write data contents into data register */
+		MAP_I2CMasterDataPut(base, *(data->msg.buf));
+		data->msg.buf++;
+
+		if (data->msg.len < 2) {
+			/* Everything has been sent, nothing to receive */
+			/* Send last byte with STOP bit */
+			MAP_I2CMasterControl(base,
+					     I2C_MASTER_CMD_BURST_SEND_FINISH);
+		} else {
+			/*
+			 * Either there is more data to be transmitted or some
+			 * data needs to be received next
+			 */
+			MAP_I2CMasterControl(base,
+					     I2C_MASTER_CMD_BURST_SEND_CONT);
+
+		}
+	} else {
+		/*
+		 * No more data needs to be sent, so follow up with
+		 * a STOP bit.
+		 * Again, there is no equivalent macro (0x00000004) so
```

Comment:
I'm not sure this comment is helpful, maybe we should remove it. There are two more places where it's used. The lines talking about "no equivalent macro".

Reply:
OK, will remove.